### ISSUE TITLE
Build number in -dev versioning

### ISSUE THREAD

**Issue by hroff-1902** [freqtrade/freqtrade#2166]
Since we do not have a build number in the -dev version, it may be quite complex to understand which particular version of freqtrade the user runs (especially with that your conda). So if we change user interface I guess it's better for a while to make those changes along with changing the version of both develop and master...

Is it possible to introduce the build number into the -dev versioning, which will change with every merge, as ccxt employs, for example?

It's not necessary for master (2019-8 is completely enough), but for develop changing every day this may be useful since `freqtrade -V` will give particular version info (even when a user has current codebase from github fetched in a dir, but installed an outdated version with that your conda, which is used instead of latest develop from the dir; that's the confusing usecase when it's complicated to find out the actual version of freqtrade that runs at the user site...)

It's not necessary to be an increasing number, it can be a short commit id (merged into develop), for example...

**Comment by xmatthias** [comment]
there's a few things to this ... it's definitely possible, however 
it would require us to:
a) open the develop branch to direct push (which is 2 clicks away - but i don't like it from a workflow perspective)
b) create a helper-api-key to allow travis / ci to push to the github repository.

I don't really like either of the two, mainly not the 2nd one since travis themselfs claim that they cannot really guarantee to protect secrets (it's always custom code that runs ...).

However, since using develop implies using a checked out git repository, we could return the commit-id instead (there's a python `git` module iirc which can do this).

This should then work for all branches, not just develop - so we can hopefully immediately see if someone is on any other branch.

Basically, the result would be `2019-8-dev_af51ff41' ... 

The downside is that this will (probably) only work when installed using `pip install -e .` - but not when intalled using `pip install .` (non linked install).

**Comment by hroff-1902** [comment]
Agree completely, thanks for detailed explanation. This requires changing the total project workflow, which in a doubtful way...

I think I have a better solution, that would allow to print in the `-V` option output the exact state of the git repo clone at the user's site (if found):

https://github.com/metaperl/sample-market-maker/blob/1a70c34ebf421f1fc2f4f2299fef730f6f1fba26/market_maker/utils/constants.py#L5-L9

This can be only run for the `-V` option, not during normal bot run.

**Comment by xmatthias** [comment]
I think to get the commit, we have 2 options:
a) use subprocess and catch ALL errors it may throw since we cannot rely on git being installed (or we're on windows ...)
b) use gitpython (see below) - which has the downside of beeing a dependency where we use 1 function from).

we could be using `gitpython` for this (`pip install gitpython`)

``` python
import git
repo = git.Repo(search_parent_directories=True)
sha = repo.head.object.hexsha
```


I'm also asking myself if we shouldn't simply use `develop` as version-string for develop versions ... they are of very limited use in the develop branch ...

**Comment by hroff-1902** [comment]
@xmatthias have a look at this: https://github.com/scikit-optimize/scikit-optimize/blob/master/versioneer.py

**Comment by xmatthias** [comment]
i've seen this before ... but trying to get this to work (copy/paste the file to freqtrade root dir) and configuring some things yields `'version': 'colorized-schema-0+562.g74a0f442.dirty'` .. which is not even the last git commit ... letalone that the name is wrong).

configuration in `setup.cfg`:
``` ini
[versioneer]
VCS = git
versionfile_source = freqtrade/__init__.py
tag_prefix =
```

Also, to use this as "single step" install we need to drop this file into the root dir (just as scikit-optimize does) - which means that we'll have a blob of 1800 lines we don't know much about, and we probably will not keep uptodate ...

in addition: there's a LOT of boilerplate that runs on top of this file:
https://github.com/scikit-optimize/scikit-optimize/blob/master/skopt/_version.py

that's too much effort for me just to determine the develop version (we'll have the release-versions done manually anyway).
